### PR TITLE
Add `maxAge` option

### LINKED ISSUE #18: Support max-age
...maxAge should be in seconds!

### PR DESCRIPTION
**Pull request by javi11**
Add a lazy max-age implementation that lacy removes expired items on get, peek, or has method calls. Hope the implementation fits on your library :).

Closes #18

### CODE REVIEW

File: test.js [hunk 1]
```diff
@@ -204,3 +221,83 @@ test('`onEviction` option method is called after `maxSize` is exceeded', t => {
 	t.is(actualValue, expectValue);
 	t.true(isCalled);
 });
+
+test('max age should remove the item on get it again', async t => {
```
Comment (sindresorhus) [inline test.js:225]: Prefix all the `maxAge` tests.
```suggestion
test('maxAge - max age should remove the item on get it again', async t => {
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by javi11** [comment]
Done ;). Thanks for the comments

**Comment by sindresorhus** [comment]
Can you fix the merge conflict?

File: index.d.ts [hunk 2]
```diff
@@ -5,6 +5,11 @@ declare namespace QuickLRU {
 		*/
 		readonly maxSize: number;
 
+		/**
+		The maximum time in milliseconds before evicting the item from the cache. Lazy expiration upon next get() or peek() of an item
```
Comment (sindresorhus) [inline index.d.ts:9]: The text here should be exactly the same as in the readme.

File: index.js [hunk 3]
```diff
@@ -6,15 +6,41 @@ class QuickLRU {
...
+		}
+
+		return item.value;
+	}
+
+	_set(key, value, expiry) {
+		if (this.maxAge > 0) {
+			this.cache.set(key, {
+				value,
+				expiry: expiry || Date.now() + this.maxAge
```
Comment (sindresorhus) [inline index.js:38]:
```suggestion
				expiry: expiry || (Date.now() + this.maxAge)
```

to make it more readable

File: index.js [hunk 4]
```diff
@@ -6,15 +6,41 @@ class QuickLRU {
...
+		this.maxAge = options.maxAge || 0;
 		this.onEviction = options.onEviction;
 		this.cache = new Map();
 		this.oldCache = new Map();
 		this._size = 0;
 	}
 
-	_set(key, value) {
-		this.cache.set(key, value);
+	_hasExpired(key, item) {
```
Comment (sindresorhus) [inline index.js:21]: `_hasExpired` makes it sound like it returns a boolean.

File: test.js [hunk 5]
```diff
@@ -204,3 +221,83 @@ test('`onEviction` option method is called after `maxSize` is exceeded', t => {
...
+	await sleep(5);
+	t.is(lru.get('1'), 'test2');
+});
+
+test('max age - once an item expires the eviction function should be called', async t => {
+	t.timeout(1000);
+	const lru = new QuickLRU({
+		maxSize: 2,
+		maxAge: 10,
+		onEviction() {
```
Comment (sindresorhus) [inline test.js:255]: Test that the arguments passed here are correct too.

File: test.js [hunk 6]
```diff
@@ -204,3 +221,83 @@ test('`onEviction` option method is called after `maxSize` is exceeded', t => {
...
+			t.pass('Test passed');
+		}
+	});
+	lru.set('1', 'test');
+	await sleep(20);
```
Comment (sindresorhus) [inline test.js:260]: Choose a larger interval. This can create fragile tests on CI.

File: test.js [hunk 7]
```diff
@@ -204,3 +221,83 @@ test('`onEviction` option method is called after `maxSize` is exceeded', t => {
...
+	await sleep(50);
+	t.is(lru.peek('1'), undefined);
+});
+
+test('max age - non recent items that are not exipred are also valid', async t => {
+	const lru = new QuickLRU({maxSize: 2, maxAge: 10});
+	lru.set('1', 'test');
+	lru.set('2', 'test2');
+	lru.set('3', 'test4');
+	await sleep(5);
```
Comment (sindresorhus) [inline test.js:285]: Too little interval. Choose larger interval for both sleep and maxAge.

**Comment by sindresorhus** [comment]
In the `has()` method. `Boolean(cache.get(...))` is not the same as `cache.has(...)`.

File: index.js (hunk 4 above)
Comment (javi11) [reply index.js]: This new name makes more sense for you, or maybe should we split the function in order to be more atomic?

**Comment by javi11** [comment]
> In the `has()` method. `Boolean(cache.get(...))` is not the same as `cache.has(...)`.

mm depending what we want if we want that has return true for items that has been expired we can remove that part, but if has will return false for expired elements, there is the need to get the element first to verify the expiration time :S

**Comment by sindresorhus** [comment]
Can you do a test where `maxAge` option is set, the value is set to `undefined` (which is different from no key being there), and using `.has()`. It looks like that will be a false-positive.

File: index.js [hunk 8]
```diff
@@ -59,15 +112,30 @@ class QuickLRU {
...
+			}
+
 			return this.cache.get(key);
 		}
 
 		if (this.oldCache.has(key)) {
+			if (this.maxAge > 0) {
+				return this._getOrDeleteIfExpired(key, this.oldCache.get(key));
+			}
+
```
Comment (sindresorhus) [inline index.js:138]: It would be great if you could go through the code and try to simplify things and DRY things up as this PR introduces a lot of boilerplate and complexity. For example, both of the instructions above are the same except for the cache, so they both could use an internal `_peek` method that accepts a cache.

File: readme.md [hunk 9]
```diff
@@ -45,6 +45,10 @@ Type: `number`
 
 The maximum number of items before evicting the least recently used items.
 
+#### maxAge
+
+If maxAge milliseconds is specified for the cache, the items will have an expiration time defined by this maxAge. Lazy expiration upon next get, peek call of the item.
```
Comment (sindresorhus) [inline readme.md:50]: I think this could be written more succinctly. You don't have to mention `maxAge` as that's clear from the context. You're also missing a type annotation.

File: test.js [hunk 10]
```diff
@@ -205,6 +222,172 @@ test('`onEviction` option method is called after `maxSize` is exceeded', t => {
 	t.true(isCalled);
 });
 
+test('max age - should remove the item on get it again', async t => {
```
Comment (sindresorhus) [inline test.js:225]:
```suggestion
test('max age - should remove the item on and it again', async t => {
```

File: test.js [hunk 11]
```diff
@@ -205,6 +222,172 @@ test('`onEviction` option method is called after `maxSize` is exceeded', t => {
...
+	const lru = new QuickLRU({maxSize: 10, maxAge: 90});
+	lru.set('1', 'test');
+	await sleep(200);
+	t.is(lru.get('1'), undefined);
+});
+
+test('max age - a non recent item can also expire', async t => {
```
Comment (sindresorhus) [inline test.js:232]:
```suggestion
test('max age - a non-recent item can also expire', async t => {
```

File: test.js [hunk 12]
```diff
@@ -205,6 +222,172 @@ test('`onEviction` option method is called after `maxSize` is exceeded', t => {
...
+	const lru = new QuickLRU({maxSize: 2, maxAge: 100});
+	lru.set('1', 'test');
+	lru.set('2', 'test2');
+	lru.set('3', 'test4');
+	await sleep(200);
+	t.is(lru.get('1'), undefined);
+});
+
+test('max age - set the item again should refresh the expiration time', async t => {
```
Comment (sindresorhus) [inline test.js:241]:
```suggestion
test('max age - setting the item again should refresh the expiration time', async t => {
```

File: test.js [hunk 13]
```diff
@@ -205,6 +222,172 @@ test('`onEviction` option method is called after `maxSize` is exceeded', t => {
...
+	const lru = new QuickLRU({maxSize: 2, maxAge: 100});
+	lru.set('1', 'test');
+	await sleep(50);
+	lru.set('1', 'test2');
+	await sleep(50);
+	t.is(lru.get('1'), 'test2');
+});
+
+test('max age - once an item expires the eviction function should be called', async t => {
```
Comment (sindresorhus) [inline test.js:250]:
```suggestion
test('max age - once an item expires, the eviction function should be called', async t => {
```

File: test.js [hunk 14]
```diff
@@ -205,6 +222,172 @@ test('`onEviction` option method is called after `maxSize` is exceeded', t => {
...
+
+	await sleep(200);
+
+	t.is(lru.get('1'), undefined);
+	t.true(isCalled);
+	t.is(actualKey, expectKey);
+	t.is(actualValue, expectValue);
+});
+
+test('max age - once an non recent item expires the eviction function should be called', async t => {
```
Comment (sindresorhus) [inline test.js:280]:
```suggestion
test('max age - once an non-recent item expires, the eviction function should be called', async t => {
```

File: test.js [hunk 15]
```diff
@@ -205,6 +222,172 @@ test('`onEviction` option method is called after `maxSize` is exceeded', t => {
...
+
+	await sleep(200);
+
+	t.is(lru.get('1'), undefined);
+	t.true(isCalled);
+	t.deepEqual(actualKeys, expectKeys);
+	t.deepEqual(actualValues, expectValues);
+});
+
+test('max age - on recise max aged items should also be evicted', async t => {
```
Comment (sindresorhus) [inline test.js:314]:
```suggestion
test('max age - on resize, max aged items should also be evicted', async t => {
```

I think you could have spent a little more time trying to reduce typos.

File: test.js [hunk 16]
```diff
@@ -205,6 +222,172 @@ test('`onEviction` option method is called after `maxSize` is exceeded', t => {
...
+
+	await sleep(200);
+
+	t.is(lru.get('1'), undefined);
+	t.true(isCalled);
+	t.deepEqual(actualKeys, expectKeys);
+	t.deepEqual(actualValues, expectValues);
+});
+
+test('max age - peek the item should also remove the item if has expired', async t => {
```
Comment (sindresorhus) [inline test.js:350]:
```suggestion
test('max age - peeking the item should also remove the item if it has expired', async t => {
```

File: index.js (hunk 8 above)
Comment (javi11) [reply index.js:138]: true, okey let me refactor it a little bit ;)

File: test.js [hunk 17]
```diff
@@ -205,6 +222,172 @@ test('`onEviction` option method is called after `maxSize` is exceeded', t => {
...
+	const lru = new QuickLRU({maxSize: 10, maxAge: 100});
+	lru.set('1', 'test');
+	await sleep(200);
+	t.is(lru.peek('1'), undefined);
+});
+
+test('max age - peek the item should also remove expired items that are not recent', async t => {
```
Comment (sindresorhus) [inline test.js:357]:
```suggestion
test('max age - peeking the item should also remove expired items that are not recent', async t => {
```

File: test.js [hunk 18]
```diff
@@ -205,6 +222,172 @@ test('`onEviction` option method is called after `maxSize` is exceeded', t => {
...
+	const lru = new QuickLRU({maxSize: 2, maxAge: 100});
+	lru.set('1', 'test');
+	lru.set('2', 'test');
+	lru.set('3', 'test');
+	await sleep(200);
+	t.is(lru.peek('1'), undefined);
+});
+
+test('max age - non recent items that are not exipred are also valid', async t => {
```
Comment (sindresorhus) [inline test.js:366]:
```suggestion
test('max age - non-recent items that are not exipred are also valid', async t => {
```

File: test.js [hunk 19]
```diff
@@ -205,6 +222,172 @@ test('`onEviction` option method is called after `maxSize` is exceeded', t => {
...
+	const lru = new QuickLRU({maxSize: 2, maxAge: 200});
+	lru.set('1', 'test');
+	lru.set('2', 'test2');
+	lru.set('3', 'test4');
+	await sleep(100);
+	t.is(lru.get('1'), 'test');
+});
+
+test('max age - has the item should also remove expired items', async t => {
```
Comment (sindresorhus) [inline test.js:375]: Not sure what you meant to write here.

File: test.js [hunk 20]
```diff
@@ -205,6 +222,172 @@ test('`onEviction` option method is called after `maxSize` is exceeded', t => {
...
+	const lru = new QuickLRU({maxSize: 2, maxAge: 100});
+	lru.set('1', 'test');
+	await sleep(200);
+	t.is(lru.has('1'), false);
+});
+
+test('max age - has the item should also remove expired items that are not recent', async t => {
```
Comment (sindresorhus) [inline test.js:382]: Same as above.

File: test.js [hunk 21]
```diff
@@ -205,6 +222,172 @@ test('`onEviction` option method is called after `maxSize` is exceeded', t => {
...
+	const lru = new QuickLRU({maxSize: 2, maxAge: 100});
+	lru.set('1', 'test');
+	lru.set('2', 'test');
+	lru.set('3', 'test');
+	await sleep(200);
+	t.is(lru.has('1'), false);
```
Comment (sindresorhus) [inline test.js:388]: Use `t.false`. Applies elsewhere too.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by javi11** [comment]
Done thanks for the comments :)

**Comment by javi11** [comment]
@sindresorhus do you see any other change that needs to be done? should I merge the commits in order to have a better history?

File: index.js [hunk 22]
```diff
@@ -96,47 +105,34 @@ class QuickLRU {
...
+			this._set(key, value, false);
 		}
 
 		return this;
 	}
 
 	has(key) {
 		if (this.maxAge > 0) {
-			return Boolean(this._getOrDeleteIfExpired(
+			const item = this.cache.get(key) || this.oldCache.get(key);
```
Comment (sindresorhus) [inline index.js:115]: This is incorrect.

From the package description:

> Inspired by the hashlru algorithm, but instead uses Map to support keys of any type, not just strings, **and values can be undefined.**

If the value is `undefined` here, it falls back to the `oldCache`, which is incorrect. Please add lots of tests for the `undefined` value case with the different methods when you fix this.

**Comment by sindresorhus** [comment]
There's a lot going on, so I'll need to do a deeper review tomorrow just to ensure there's no faulty logic.

In the meantime, it would be great if you could also review your own changes in depth. When doing a large addition like this, it's easy to miss stuff. You could also try to add more tests. There are never too many tests.

File: index.d.ts [hunk 23]
```diff
@@ -5,6 +5,13 @@ declare namespace QuickLRU {
 		*/
 		readonly maxSize: number;
 
+		/**
+		The maximum number of milliseconds an item should remain in cache.
+
+		Lazy expiration upon the next `.get()`, `.peek()`, or `.has()` call.
```
Comment (sindresorhus) [inline index.d.ts:11]: You need to document the default.

**Comment by sindresorhus** [comment]
You missed the last part of https://github.com/sindresorhus/quick-lru/pull/27#discussion_r530852496

File: index.js (hunk 22 above)
Comment (javi11) [reply index.js:115]: Yep you are right, let me add more test for this cases ;)

**Comment by javi11** [comment]
Hi @sindresorhus I did some refactor of the code and add more testing, let me know if you find some other edge case, I will also take time to look for more edge cases.

Thanks for the help ;)

File: index.d.ts [hunk 24]
```diff
@@ -22,20 +31,14 @@ declare class QuickLRU<KeyType, ValueType> implements Iterable<[KeyType, ValueTy
...
 	```
 	import QuickLRU = require('quick-lru');
-
 	const lru = new QuickLRU({maxSize: 1000});
-
 	lru.set('🦄', '🌈');
-
 	lru.has('🦄');
 	//=> true
-
```
Comment (sindresorhus) [inline index.d.ts:38]: Don't do unrelated changes. Applies in very many places.

File: index.d.ts [hunk 25]
```diff
@@ -6,8 +6,17 @@ declare namespace QuickLRU {
 		readonly maxSize: number;
 
 		/**
-		Called right before an item is evicted from the cache.
+		The maximum number of milliseconds an item should remain in cache.
+		By default maxAge will be 0 which means that items will never expires.
```
Comment (sindresorhus) [inline index.d.ts:10]:
```suggestion

		By default, `maxAge` will be `0`, which means that items will never expire.
```

File: index.d.ts [hunk 26]
```diff
@@ -6,8 +6,17 @@ declare namespace QuickLRU {
...
+
+		Lazy expiration upon the next `write` or `read` call.
 
+		Individual expiration of an item can be specified by the `set(key, value, expiry)` method.
```
Comment (sindresorhus) [inline index.d.ts:14]:
```suggestion
		Individual expiration of an item can be specified with the `set(key, value, expiry)` method.
```

File: index.js [hunk 27]
```diff
@@ -35,40 +69,77 @@ class QuickLRU {
...
+			const item = this.oldCache.get(key);
+			if (this._deleteIfExpired(key, item) === false) {
+				this._moveToRecent(key, item);
+				return item.value;
+			}
 		}
 	}
 
-	set(key, value) {
+	set(key, value, expiry = this.maxAge > 0 ? Date.now() + this.maxAge : null) {
```
Comment (sindresorhus) [inline index.js:113]: Use `undefined`, not `null`.

File: readme.md [hunk 28]
```diff
@@ -60,7 +72,7 @@ The instance is an [`Iterable`](https://developer.mozilla.org/en/docs/Web/JavaSc
 
 Both `key` and `value` can be of any type.
 
-#### .set(key, value)
+#### .set(key, value, expiry)
```
Comment (sindresorhus) [inline readme.md:74]: The `expiry` parameter needs to be documented.

File: test.js [hunk 29]
```diff
@@ -9,12 +9,23 @@ const lruWithDuplicates = () => {
...
+		setTimeout(resolve, ms);
+	});
+
 test('main', t => {
 	t.throws(() => {
 		new QuickLRU(); // eslint-disable-line no-new
 	}, /maxSize/);
 });
 
+test('wrong max age', t => {
```
Comment (sindresorhus) [inline test.js:23]:
```suggestion
test('max age - incorrect value', t => {
```

File: test.js [hunk 30]
```diff
@@ -205,6 +216,362 @@ test('`onEviction` option method is called after `maxSize` is exceeded', t => {
...
+	lru.set('1', undefined);
+	lru.set('2', 'test2');
+	lru.set('3', 'test3');
+	await sleep(200);
+	lru.set('4', 'loco');
+
+	t.deepEqual([...lru.keys()].sort(), ['4']);
+});
+
+test('max age - keys() should return an empty list if all items has expired', async t => {
```
Comment (sindresorhus) [inline test.js:452]:
```suggestion
test('max age - `.keys()` should return an empty list if all items has expired', async t => {
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by javi11** [comment]
@sindresorhus I did the mentioned changes let me know if you need anything else. Happy new year BTW.

File: index.d.ts [hunk 31]
```diff
@@ -1,5 +1,14 @@
 declare namespace QuickLRU {
 	interface Options<KeyType, ValueType> {
+		/**
```
Comment (sindresorhus) [inline index.d.ts:3]: This needs an actual description of the option first.

I also noticed we don't explain in what unit the max age.

File: index.d.ts [hunk 32]
```diff
@@ -49,7 +58,7 @@ declare class QuickLRU<KeyType, ValueType> implements Iterable<[KeyType, ValueTy
 
 	@returns The list instance.
 	*/
-	set(key: KeyType, value: ValueType): this;
+	set(key: KeyType, value: ValueType, expiry?: number): this;
```
Comment (sindresorhus) [inline index.d.ts:61]: This should be `maxAge` for consistency and needs to be documented.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
Don't forget that index.d.ts and readme should be in sync.

File: index.d.ts (hunk 32 above)
Comment (javi11) [reply index.d.ts]: Done and documented let me know if its' fine for you

File: index.d.ts (hunk 31 above)
Comment (javi11) [reply index.d.ts:3]: mm I don't follow you with the "I also noticed we don't explain in what unit the max age." what do you mean?

File: index.d.ts [hunk 33]
```diff
@@ -47,9 +57,13 @@ declare class QuickLRU<KeyType, ValueType> implements Iterable<[KeyType, ValueTy
...
 	Set an item.
 
+	Individual expiration of an item can be specified with the `maxAge` parameter.
+	If not specified, the global `maxAge` value will be used in case that is specified on the constructor,
+	else the item will not have expiration.
+
 	@returns The list instance.
 	*/
-	set(key: KeyType, value: ValueType): this;
+	set(key: KeyType, value: ValueType, magAge?: number): this;
```
Comment (papb) [inline index.d.ts:66]:
```suggestion
	set(key: KeyType, value: ValueType, maxAge?: number): this;
```

File: readme.md [hunk 34]
```diff
@@ -45,6 +45,17 @@ Type: `number`
 
 The maximum number of items before evicting the least recently used items.
 
+#### maxAge
+
+Type: `number`
+
+The maximum number of milliseconds an item should remain in cache.
+By default maxAge will be 0 which means that items will never expires.
```
Comment (papb) [inline readme.md:53]:
```suggestion
By default maxAge will be 0, which means that items will never expire.
```

File: readme.md [hunk 35]
```diff
@@ -45,6 +45,17 @@ Type: `number`
...
+
+Lazy expiration upon the next `write` or `read` call.
```
Comment (papb) [inline readme.md:55]:
```suggestion
Lazy expiration happens upon the next `write` or `read` call.
```

File: readme.md [hunk 36]
```diff
@@ -60,10 +71,12 @@ The instance is an [`Iterable`](https://developer.mozilla.org/en/docs/Web/JavaSc
 
 Both `key` and `value` can be of any type.
 
-#### .set(key, value)
+#### .set(key, value, maxAge)
 
 Set an item. Returns the instance.
 
+Individual expiration of an item can be specified with the `maxAge` parameter. If not specified, the global `maxAge` value will be used in case that is specified on the constructor, else the item will not have expiration.
```
Comment (papb) [inline readme.md:78]:
```suggestion
Individual expiration of an item can be specified with the `maxAge` parameter. If not specified, the global `maxAge` value will be used in case it is specified on the constructor, otherwise the item will never expire.
```

**Review by papb — CHANGES_REQUESTED** [review]

File: index.d.ts [hunk 37]
```diff
@@ -47,9 +57,13 @@ declare class QuickLRU<KeyType, ValueType> implements Iterable<[KeyType, ValueTy
...
 	Set an item.
 
+	Individual expiration of an item can be specified with the `maxAge` parameter.
+	If not specified, the global `maxAge` value will be used in case that is specified on the constructor,
+	else the item will not have expiration.
+
 	@returns The list instance.
 	*/
-	set(key: KeyType, value: ValueType): this;
+	set(key: KeyType, value: ValueType, maxAge?: number): this;
```
Comment (sindresorhus) [inline index.d.ts:66]: Looking at the callsite of this, I don't think `set('foo', 1, 100)` is very readable. It's not clear to the person reading it what the last number is. I think it should be an options-object instead. Sorry for not catching this earlier.

File: index.d.ts [hunk 38]
```diff
@@ -47,9 +57,13 @@ declare class QuickLRU<KeyType, ValueType> implements Iterable<[KeyType, ValueTy
 	/**
 	Set an item.
 
+	Individual expiration of an item can be specified with the `maxAge` parameter.
+	If not specified, the global `maxAge` value will be used in case that is specified on the constructor,
+	else the item will not have expiration.
```
Comment (sindresorhus) [inline index.d.ts:62]: Don't hard-wrap.

File: index.d.ts [hunk 39]
```diff
@@ -1,5 +1,15 @@
 declare namespace QuickLRU {
 	interface Options<KeyType, ValueType> {
+		/**
+		The maximum number of milliseconds an item should remain in cache.
+
+		By default maxAge will be 0 which means that items will never expires.
+		Lazy expiration upon the next `write` or `read` call.
```
Comment (sindresorhus) [inline index.d.ts:7]:
```suggestion
		Lazy expiration upon the next write or read call.
```

File: index.d.ts [hunk 40]
```diff
@@ -1,5 +1,15 @@
 declare namespace QuickLRU {
 	interface Options<KeyType, ValueType> {
+		/**
+		The maximum number of milliseconds an item should remain in cache.
+
+		By default maxAge will be 0 which means that items will never expires.
```
Comment (sindresorhus) [inline index.d.ts:6]:
```suggestion
		By default, `maxAge` will be `0` which means that items will never expires.
```

File: index.d.ts [hunk 41]
```diff
@@ -1,5 +1,15 @@
 declare namespace QuickLRU {
 	interface Options<KeyType, ValueType> {
+		/**
+		The maximum number of milliseconds an item should remain in cache.
+
```
Comment (sindresorhus) [inline index.d.ts:5]:
```suggestion

@default 0
```

File: index.d.ts (hunk 40 above)
Comment (sindresorhus) [reply index.d.ts:6]: It's kinda weird that `0` means to never expire. I think `Infinity` would be a better default to indicate that. `0` makes it sound like it will expire right away.

File: readme.md [hunk 42]
```diff
@@ -45,6 +45,17 @@ Type: `number`
 
 The maximum number of items before evicting the least recently used items.
 
+#### maxAge
+
+Type: `number`
```
Comment (sindresorhus) [inline readme.md:50]:
```suggestion
Type: `number`\
Default: `0`
```

File: readme.md [hunk 43]
```diff
@@ -60,10 +71,12 @@ The instance is an [`Iterable`](https://developer.mozilla.org/en/docs/Web/JavaSc
 
 Both `key` and `value` can be of any type.
 
-#### .set(key, value)
+#### .set(key, value, maxAge)
```
Comment (sindresorhus) [inline readme.md:74]:
```suggestion
#### .set(key, value, maxAge?)
```

File: test.js [hunk 44]
```diff
@@ -205,6 +216,369 @@ test('`onEviction` option method is called after `maxSize` is exceeded', t => {
...
+	lru.set('1', undefined);
+	lru.set('2', 'test2');
+	lru.set('3', 'test3');
+	await sleep(200);
+	lru.set('5', 'loco');
+
+	t.deepEqual([...lru.values()].sort(), ['loco']);
+});
+
+test('max age - `entriesDescending() should not return expired entries', async t => {
```
Comment (sindresorhus) [inline test.js:490]:
```suggestion
test('max age - `entriesDescending()` should not return expired entries', async t => {
```

Same in many other places. "Method" calls should be wrapped in backticks.

**Comment by javi11** [comment]
Done, I called the set option maxAge to options = { maxAge } and set maxAge globaly by default Infinity instead of 0 that sure it makes more sense ;)

File: index.d.ts [hunk 45]
```diff
@@ -1,5 +1,16 @@
 declare namespace QuickLRU {
 	interface Options<KeyType, ValueType> {
+		/**
+		The maximum number of milliseconds an item should remain in cache.
```
Comment (sindresorhus) [inline index.d.ts:4]:
```suggestion
		The maximum number of milliseconds an item should remain in the cache.
```

File: index.d.ts [hunk 46]
```diff
@@ -1,5 +1,16 @@
...
+
+		@default Infinity
```
Comment (sindresorhus) [inline index.d.ts:6]:
```suggestion
		@default Infinity

```

File: index.d.ts [hunk 47]
```diff
@@ -1,5 +1,16 @@
...
+		By default, `maxAge` will be `Infinity` which means that items will never expires.
```
Comment (sindresorhus) [inline index.d.ts:7]:
```suggestion
		By default, `maxAge` will be `Infinity`, which means that items will never expire.
```

File: index.d.ts [hunk 48]
```diff
@@ -45,11 +56,13 @@ declare class QuickLRU<KeyType, ValueType> implements Iterable<[KeyType, ValueTy
 	[Symbol.iterator](): IterableIterator<[KeyType, ValueType]>;
 
 	/**
-	Set an item.
+	Set an item. Returns the instance.
+
+	Individual expiration of an item can be specified with the `maxAge` option. If not specified, the global `maxAge` value will be used in case it is specified on the constructor, otherwise the item will never expire.
```
Comment (sindresorhus) [inline index.d.ts:61]:
```suggestion
	Individual expiration of an item can be specified with the `maxAge` option. If not specified, the global `maxAge` value will be used in case it is specified in the constructor, otherwise the item will never expire.
```

File: index.d.ts [hunk 49]
```diff
@@ -45,11 +56,13 @@ declare class QuickLRU<KeyType, ValueType> implements Iterable<[KeyType, ValueTy
...
 
 	@returns The list instance.
 	*/
-	set(key: KeyType, value: ValueType): this;
+	set(key: KeyType, value: ValueType, options?: { maxAge?: number }): this;
```
Comment (sindresorhus) [inline index.d.ts:65]:
```suggestion
	set(key: KeyType, value: ValueType, options?: {maxAge?: number}): this;
```

File: readme.md [hunk 50]
```diff
@@ -60,10 +72,12 @@ The instance is an [`Iterable`](https://developer.mozilla.org/en/docs/Web/JavaSc
 
 Both `key` and `value` can be of any type.
 
-#### .set(key, value)
+#### .set(key, value, options?)
 
 Set an item. Returns the instance.
 
+Individual expiration of an item can be specified with the `maxAge` parameter. If not specified, the global `maxAge` value will be used in case it is specified on the constructor, otherwise the item will never expire.
```
Comment (sindresorhus) [inline readme.md:79]: Outdated docs.

File: readme.md [hunk 51]
```diff
@@ -45,6 +45,18 @@ Type: `number`
 
 The maximum number of items before evicting the least recently used items.
 
+#### maxAge
+
+Type: `number`
```
Comment (sindresorhus) [inline readme.md:50]:
```suggestion
Type: `number`\
```
